cpp/benchmarks/streaming/ndsh/q18.cpp

```diff
@@ -0,0 +1,1455 @@
+        nodes.push_back(final_groupby_filter_lineitem(
+            ctx, concatenated, filtered_local, quantity_threshold
+        ));
+
+        // All-gather the TINY filtered result (~57K orderkeys at SF1000)
+        auto gathered = ctx->create_channel();
+        nodes.push_back(allgather_table(
+            ctx,
+            filtered_local,
+            gathered,
+            rapidsmpf::OpID{static_cast<rapidsmpf::OpID>(base_tag + 1)}
+        ));
+        to_collect = gathered;
+    }
+
+    // Collect result using pull_from_channel (safe coroutine pattern)
+    std::vector<rapidsmpf::streaming::Message> result_messages;
+    nodes.push_back(
+        rapidsmpf::streaming::node::pull_from_channel(ctx, to_collect, result_messages)
+    );
+
+    // Run pipeline
+    rapidsmpf::streaming::run_streaming_pipeline(std::move(nodes));
+
+    // Extract result from collected messages
+    std::unique_ptr<cudf::table> result;
+    if (!result_messages.empty()) {
+        auto chunk = rapidsmpf::ndsh::to_device(
+            ctx, result_messages[0].release<rapidsmpf::streaming::TableChunk>()
+        );
+        auto stream = chunk.stream();
+        auto table_view = chunk.table_view();
+
+        // Extract just the orderkey column (column 0)
+        // The filtered result has (l_orderkey, sum_quantity), we only need l_orderkey
+        std::vector<std::unique_ptr<cudf::column>> cols;
+        cols.push_back(
+            std::make_unique<cudf::column>(
+                table_view.column(0), stream, ctx->br()->device_mr()
+            )
+        );
+        stream.synchronize();
+        result = std::make_unique<cudf::table>(std::move(cols));
+    }
+
+    ctx->comm()->logger().print(
+        "Phase 1 complete: ", result ? result->num_rows() : 0, " qualifying orderkeys"
+    );
+
+    return result;
+}
+
+// ============================================================================
+// Phase 2: Pre-filter Pipeline
+// ============================================================================
+
+/**
+ * @brief Pre-filter table by qualifying orderkeys using semi-join.
+ *
+ * @param qualifying_orderkeys Table with single column of qualifying l_orderkey values.
+ * @param key_column_idx Which column in input chunks to match against orderkeys.
+ */
+rapidsmpf::streaming::Node prefilter_by_orderkeys(
+    std::shared_ptr<rapidsmpf::streaming::Context> ctx,
+    std::shared_ptr<rapidsmpf::streaming::Channel> ch_in,
+    std::shared_ptr<rapidsmpf::streaming::Channel> ch_out,
+    std::shared_ptr<cudf::table> qualifying_orderkeys,
+    cudf::size_type key_column_idx
+) {
+    rapidsmpf::streaming::ShutdownAtExit c{ch_in, ch_out};
+    std::uint64_t sequence = 0;
+
+    // Build filtered_join for semi-join (orderkeys is the "right"/build side)
+    auto joiner = cudf::filtered_join(
+        qualifying_orderkeys->view(),
+        cudf::null_equality::UNEQUAL,
+        cudf::set_as_build_table::RIGHT,
+        cudf::get_default_stream()
```

Comment:
I don't think we want `cudf::get_default_stream()` here. Perhaps we can use whatever stream `qualifying_orderkeys` came in on?